cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java

```diff
@@ -190,9 +211,61 @@ public void beforeSubmit(MapReduceContext context) throws Exception {
     }
 
     job.setMapperClass(ETLMapper.class);
-    job.setNumReduceTasks(0);
+    Set<StageInfo> aggregators = phaseSpec.getPhase().getStagesOfType(BatchAggregator.PLUGIN_TYPE);
+    if (!aggregators.isEmpty()) {
+      job.setReducerClass(ETLReducer.class);
+      String aggregatorName = aggregators.iterator().next().getName();
+      BatchAggregator aggregator = pluginInstantiator.newPluginInstance(aggregatorName);
+      MapReduceAggregatorContext aggregatorContext =
+        new MapReduceAggregatorContext(context, mrMetrics,
+                                       new DatasetContextLookupProvider(context),
+                                       aggregatorName, context.getRuntimeArguments());
+      aggregator.prepareRun(aggregatorContext);
+
+      if (aggregatorContext.getNumPartitions() != null) {
+        job.setNumReduceTasks(aggregatorContext.getNumPartitions());
+      }
+      // if the plugin sets the output key and value class directly, trust them
+      Class<?> outputKeyClass = aggregatorContext.getGroupKeyClass();
+      Class<?> outputValClass = aggregatorContext.getGroupValueClass();
+      // otherwise, derive it from the plugin's parameters
+      if (outputKeyClass == null) {
+        outputKeyClass = TypeChecker.getGroupKeyClass(aggregator);
+      }
+      if (outputValClass == null) {
+        outputValClass = TypeChecker.getGroupValueClass(aggregator);
+      }
+      hConf.set(GROUP_KEY_CLASS, outputKeyClass.getName());
+      hConf.set(GROUP_VAL_CLASS, outputValClass.getName());
+      // in case the classes are not a WritableComparable, but is some common type we support
+      // for example, a String or a StructuredRecord
+      WritableConversion writableConversion = WritableConversions.getConversion(outputKeyClass.getName());
+      if (writableConversion != null) {
```

Comment:
Will we hit the case where writableConversion is null? Some class that is provided by the user that we don't support in WritableConversions?